### PR TITLE
WIP: Add t5 adapter

### PR DESCRIPTION
**Pull request by AmirAktify**
Followed the pattern of Bart to add adapters to T5. One change is that whereas Bart has separate classes for encoder and decoder, T5 does not. So I am using the `is_decoder` for changes between encoder and decoder classes, such as adding cross_attention adapters and adding invertible adapters.

I'm working on some testing.

### CODE REVIEW

**Comment by denizbeser** [comment]
@AmirAktify thanks for working on this! My research team was looking to use adapters with T5 and we found this branch you've been working on. Do you think this is in a usable stage? Thanks!

**Comment by AmirAktify** [comment]
> @AmirAktify thanks for working on this! My research team was looking to use adapters with T5 and we found this branch you've been working on. Do you think this is in a usable stage? Thanks!

Hi Deniz. I was running some tests yesterday with sequence classification with T5ForConditionalGeneration, and I ran into some minor errors with the base model. Lemme try and debug those today, but I am hopefully very close.

**Comment by denizbeser** [comment]
> > @AmirAktify thanks for working on this! My research team was looking to use adapters with T5 and we found this branch you've been working on. Do you think this is in a usable stage? Thanks!
> 
> Hi Deniz. I was running some tests yesterday with sequence classification with T5ForConditionalGeneration, and I ran into some minor errors with the base model. Lemme try and debug those today, but I am hopefully very close.

@AmirAktify That's great to hear. Thanks for letting me know so quickly. Good luck!

**Review by calpt — COMMENTED** [review]
Looks good overall so far 👍. I've left some comments mainly on the areas where you had to modify the existing implementation.

Haven't looked into all the details yet, so this is more a rough review of the main implementation parts.

File: src/transformers/adapters/model_mixin.py [hunk 1]
```diff
@@ -88,6 +107,55 @@ def invertible_adapters_forward(self, hidden_states, rev=False):
         return hidden_states
 
 
+class EncoderRestrictedInvertibleAdaptersMixin(InvertibleAdaptersMixin):
```
[the remark this answers is not in the document]
Comment (AmirAktify) [reply src/transformers/adapters/model_mixin.py:110]: Yes. I should / can get rid of this now. Sorry about that.

File: src/transformers/adapters/model_mixin.py [hunk 2]
```diff
@@ -73,12 +86,18 @@ def get_invertible_adapter(self):
         return None
 
     def enable_invertible_adapters(self, adapter_names):
+        if self.disable_for_decoder:
+            return None
```
[the remark this answers is not in the document]
Comment (AmirAktify) [reply src/transformers/adapters/model_mixin.py:90]: Made a note of it, I'll comment done when I do a round of fixes this weekend.

File: src/transformers/adapters/model_mixin.py [hunk 3]
```diff
@@ -27,9 +27,16 @@
...
     """Mixin for Transformer models adding invertible adapters."""
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self.invertible_adapters = nn.ModuleDict(dict())
+    def __init__(self, config, *args, **kwargs):
+        # This assumes invertible adapters are never added on decoders.
+        self.disable_for_decoder = False
+        if config.is_decoder and not config.is_encoder_decoder:
+            self.disable_for_decoder = True
```
[the remark this answers is not in the document]
Comment (AmirAktify) [reply src/transformers/adapters/model_mixin.py:34]: That's a great idea. I'll do it.

File: src/transformers/adapters/layer.py [hunk 4]
```diff
@@ -455,12 +455,14 @@ def adapters_forward(self, hidden_states, input_tensor, **kwargs):
 
             last_config = self.config.adapters.get(adapter_setup.last())
             if last_config["original_ln_after"]:
-                if self.layer_norm:
+                # HACK for T5 adapters. We do not want a norm applied on adapter output,
+                # but cannot clobber the layer_norm in the base model to none.
+                if self.layer_norm and getattr(self, "allow_adapter_layer_norm", True):
```
[the remark this answers is not in the document]
Comment (AmirAktify) [reply src/transformers/adapters/layer.py:460]: Wonderful. I saw you created the branch. Are you going to merge that in to master, or do you want me to rebase my branch onto it? Whichever you'd prefer.

File: src/transformers/adapters/model_mixin.py [hunk 5]
```diff
@@ -535,42 +593,13 @@ def pre_transformer_forward(self, **kwargs):
 
 
 @inherit_doc
-class ModelWithHeadsAdaptersMixin(ModelAdaptersMixin):
-    """Mixin adding support for loading/ saving adapters to transformer models with head(s)."""
+class ModelWithHeadsAndNoBaseAdaptersMixin(ModelAdaptersMixin):
+    """Mixin adding support for loading/ saving adapters to transformer models with head(s), and no base model."""
```
[the remark this answers is not in the document]
Comment (AmirAktify) [reply src/transformers/adapters/model_mixin.py:597]: There's just a nasty infinite recursion that's created by functions patterned like this in ModelWithHeadsAdaptersMixin where the `base_model==model` is true, as you can see below:

```
    def add_adapter(self, adapter_name: str, config=None, overwrite_ok: bool = False):
        self.base_model.add_adapter(adapter_name, config, overwrite_ok=overwrite_ok)
```


There are definitely alternative workarounds where I could explicitly override all of these broken functions. It felt easier for development purposes to create an alternate class `ModelWithHeadsAdaptersMixin` that didn't come with these infinite recursion functions, but had all the head saving / loading mechanisms I wanted.

Would you like me to simply override all of the infinite recursion functions by creating a `T5ModelWithHeadsAdaptersMixin` that extends `ModelWithHeadsAdaptersMixin`? That would also work, and not restructure / split your main adapter classes. The only thing is it will break / need extension if a new infinite recursion function is added to `ModelWithHeadsAdaptersMixin` later. I'll do any this or any other alternate solutions you'd like, you know your code base best.

[the remark this answers is not in the document]
Comment (calpt) [reply src/transformers/adapters/model_mixin.py:597]: Ah, I see, I totally didn't realize the infinite recursion when looking over the code. I think both of your solution would work. Maybe as an alternative, a simple check within the methods in question would also suffice, e.g. like:
```python
def _add_adapter(self, adapter_name):
    if self.base_model is self:
        super()._add_adapter(adapter_name)
    else:
        self.base_model._add_adapter(adapter_name)
```
That wouldn't require any changes to the overall class structure.

[the remark this answers is not in the document]
Comment (AmirAktify) [reply src/transformers/adapters/model_mixin.py:597]: The super class of `ModelWithHeads`, namely `ModelAdaptersMixin`, doesn't implement `_add_adapter` though. so this would just cause a pass? No? 

Doing a T5ModelWithHeadsAdaptersMixin should also avoid changing the class structure.

[the remark this answers is not in the document]
Comment (calpt) [reply src/transformers/adapters/model_mixin.py:597]: It does actually call the implementation in `T5ModelAdaptersMixin` but as this functionality is dependent on the inheritance order in `T5ForConditionalGeneration`, after thinking about it again, I agree that adding a `T5ModelWithHeadsAdaptersMixin` probably would be the best solution.

File: src/transformers/adapters/layer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (calpt) [reply src/transformers/adapters/layer.py:460]: Is merged into master now.

[the remark this answers is not in the document]
Comment (AmirAktify) [reply src/transformers/adapters/layer.py:460]: Got caught up in a bunch of things at work last few weeks, and finally got those out of my way. This weekend I'll incorporate your changes to get this wound up. Sorry for being so slow!

File: src/transformers/adapters/model_mixin.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AmirAktify) [reply src/transformers/adapters/model_mixin.py:90]: Added exception raising.

File: src/transformers/adapters/model_mixin.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AmirAktify) [reply src/transformers/adapters/model_mixin.py:34]: Done, seems to work now.

File: src/transformers/adapters/model_mixin.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AmirAktify) [reply src/transformers/adapters/model_mixin.py]: Looking through the code more, I've reconsidered that your initial suggestion to check for self and pass to the super class is probably better. It does indeed rely on the inheritance order.......but its going to be less complex and more readable code.

[the remark this answers is not in the document]
Comment (AmirAktify) [reply src/transformers/adapters/model_mixin.py]: Thank you for the suggestion with doing the check on base model. It was very fast, and means I don't need to refactor the model structure much. That's done now.

**Comment by AmirAktify** [comment]
Addressed the first round of comments @calpt had, and the model is still training / resolving fine on IMDB. Going through fixing the unit tests now. 

One slight issue I'm seeing is that the tests assume the underlying huggingface has `AutoModelForSequenceClassification` support. T5 does not have this in the base `huggingface` currently (https://huggingface.co/transformers/model_doc/auto.html#tfautomodelforsequenceclassification), so I may just have to disable it from those tests.

**Comment by calpt** [comment]
@AmirAktify regarding the unit tests: I've done some refactoring of the test classes together with the `EncoderDecoderModel` implementation in #222 for similar reasons (e.g. `EncoderDecoderModel` also does not support `AutoModelForSequenceClassification`). Maybe we could sync this branch with that branch so the refactoring isn't done twice? (#222 should be ready for merge soon, but you could merge it to your branch already.)

File: src/transformers/adapters/model_mixin.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (calpt) [reply src/transformers/adapters/model_mixin.py]: > Looking through the code more, I've reconsidered that your initial suggestion to check for self and pass to the super class is probably better. It does indeed rely on the inheritance order.......but its going to be less complex and more readable code.

Ok. Please try to document this mechanism and the dependence on inheritance order somewhere in code using comment(s) to make sure it's still understandable in the future. Thanks! :)

**Comment by calpt** [comment]
@AmirAktify the mentioned pull request (#222) is now merged into master. I see there are still some tests failing here. Is there something else I can help with to complete this PR soon?

**Comment by AmirAktify** [comment]
> @AmirAktify regarding the unit tests: I've done some refactoring of the test classes together with the `EncoderDecoderModel` implementation in #222 for similar reasons (e.g. `EncoderDecoderModel` also does not support `AutoModelForSequenceClassification`). Maybe we could sync this branch with that branch so the refactoring isn't done twice? (#222 should be ready for merge soon, but you could merge it to your branch already.)

Hi clifton. Sorry I forgot to respond at the time you mentioned, but I had resynced it up before you merged to master after you pointed this out to me.

**Comment by AmirAktify** [comment]
> @AmirAktify the mentioned pull request (#222) is now merged into master. I see there are still some tests failing here. Is there something else I can help with to complete this PR soon?

Hi Clifton. I finally got all the unit tests passing. I had to add in a `delete_adapters` method which had come in recent pushes to master, and I had to make sure `active_adapters` was being passed properly through kwargs to the underlying layers to finally get unit tests working. 
I also checked that I'm still getting about 93% accuracy on T5ForConditionalGeneration on the IMDB task with 1.8 million trainable parameters, so th

Does your team have capacity by any chance to finish off the PR? I'm just running a little ragged at work, and I've probably almost taken more of your time with questions than it would have to wind it up yourself. I apologize for how long its been taking, and how slow I've been to do this. It's been a fun educational project for me though.

**Comment by AmirAktify** [comment]
> @AmirAktify thanks for working on this! My research team was looking to use adapters with T5 and we found this branch you've been working on. Do you think this is in a usable stage? Thanks!

I think its in a useable state in this point, finally....

**Comment by hSterz** [comment]
Hi @AmirAktify, if you grant me access to your branch I can work on finishing the Pull Request

**Comment by AmirAktify** [comment]
> Hi @AmirAktify, if you grant me access to your branch I can work on finishing the Pull Request

Done, thank you. `src/test_t5_with_adapters_on_imdb.ipynb` is what I've been using to confirm that the adapter is working / converging. I've been focusing on `T5ForConditionalGeneration`. Unit tests seem to be passing currently. Let me know if you have any questions.

File: src/transformers/adapters/models/t5.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+                if hasattr(adapt, name):
+                    if idx not in return_adapters:
+                        return_adapters[idx] = {}
+                    return_adapters[idx][key] = getattr(adapt, name)
+
+        return return_adapters
+
+
+class T5ModelHeadsMixin(ModelWithFlexibleHeadsAdaptersMixin):
+    """Adds flexible heads to a GPT-2 model."""
```
Comment (AmirAktify) [inline src/transformers/adapters/models/t5.py:227]: "Adds flexible heads to a T5 model"?

Comment (hSterz) [reply src/transformers/adapters/models/t5.py]: Yes, of course

**Comment by AmirAktify** [comment]
Thanks for finishing this PR btw! I hope it wasn't quite as bad as just doing it from scratch.

File: tests/test_pipelines_feature_extraction.py [hunk 7]
```diff
@@ -33,7 +33,215 @@ def test_small_model_pt(self):
...
...[-0.187, -1.277, 0.849, -0.439, -0.967, -1.347, 1.063, 0.469, 1.086, -1.253, 0.349, 0.057, 1.031, -1.903, -0.432, -1.377, 0.379, 0.733, -1.043, 1.307, 0.865, 0.229, 1.373, 1.671, -0.285, 0.599, -1.418, -1.179, -0.369, 1.039, -0.705, 1.082], [-1.735, 1.102, 0.398, -0.245, 1.452, 0.46, -1.734, -0.746, 1.831, 0.562, 1.464, -0.342, -0.619, -0.455, 0.127, -1.209, -0.686, -0.395, -0.316, 2.467, -0.379, 0.328, 0.639, 0.4, -1.097, -0.096, 0.397, -0.806, -1.621, 1.127, -0.345, 0.074], [0.296, -0.638, 1.938, -0.151, -1.19, 1.445, 1.318, 0.711, -0.125, 0.127, -2.179, 0.481, -1.019, 1.178, 0.318, 1.858, -1.646, 0.185, -0.072, -0.979, 0.82, -1.374, 0.836, -1.019, 0.043, -0.156, -0.095, 0.641, -0.195, -0.076, -1.554, 0.275], [-0.266, 0.971, 0.745, -0.37, 1.42, -0.5, -0.53, 0.061, 1.311, -0.1, 1.796, 0.53, -0.739, -0.325, 0.28, -1.72, 0.382, -1.118, 0.442, 1.84, -2.497, 1.003, -0.788, -0.224, -0.604, -1.259, -0.475, 1.18, -1.356, 0.695, 0.201, 0.016], [-0.618, -1.495, -0.67, -0.106, -1.265, -0.51, -1.752, 1.018, 0.674, 0.181, 0.297, 0.479, -0.185, 0.081, -2.44, -0.239, 1.081, -1.38, 0.679, 0.878, 1.336, -1.347, 0.969, -0.847, 0.293, 0.476, 1.647, -0.641, 0.66, 1.236, 0.761, 0.751]]])  # fmt: skip
```
Comment (calpt) [inline tests/test_pipelines_feature_extraction.py:36]: Please revert all changes in this file & upgrade to latest black version (`black==21.4b0`) to avoid reformatting :)

File: tests/test_adapter.py [hunk 8]
```diff
@@ -324,3 +325,26 @@ def forward_pre_hook(module, input):
...
+
+
+@require_torch
+class T5AdapterTest(
+    AdapterModelTestMixin,
+    AdapterFusionModelTestMixin,
+    PredictionHeadModelTestMixin,
+    AdapterTestBase,
+    unittest.TestCase,
+):
```
Comment (calpt) [inline tests/test_adapter.py:337]: It seems like T5 is a "full" adapter implementation (w. flex heads, parallel block, ...), so, compared to other models such as BERT & GPT-2, some additional test mixins might be useful:
- AdapterTrainingTestMixin
- ParallelAdapterInferenceTestMixin
- ParallelTrainingMixin

File: tests/test_adapter.py [hunk 9]
```diff
@@ -324,3 +325,26 @@ def forward_pre_hook(module, input):
 
         self.assertEqual((1, 128, model.config.decoder.vocab_size), out[0].shape)
         self.assertEqual(2, calls)
+
+
+@require_torch
+class T5AdapterTest(
```
Comment (calpt) [inline tests/test_adapter.py:331]: Since T5 supports flex heads & conversion `ModelClassConversionTestMixin` should be implemented (see `GPT2ClassConversionTest` for an example).

**Review by calpt — CHANGES_REQUESTED** [review]
Looks nice overall :)

Besides the line-specific comments, two general points from the [implementation guide](https://github.com/Adapter-Hub/adapter-transformers/blob/master/adding_adapters_to_a_model.md) document still seem to be missing (from section _Testing_):

>  - Insert test_modeling_<model_type> into the list of tested modules in utils/run_tests.py.
>  - Append <model_type> to the list in check_adapters.py.

**Comment by kanseaveg** [comment]
nice works!
